### PR TITLE
Chain 265/make sure replayed requests lead to same output

### PR DESCRIPTION
**Pull request by imykolenko**

### CODE REVIEW

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/OrderBook.kt [hunk 1]
```diff
@@ -150,7 +155,7 @@ class OrderBookLevel(val levelIx: Int, var side: BookSide, val price: BigDecimal
             }
         }
         // remove consumed orders
-        orderHead = ix
+        orderHead = ix // FIXME should also reset consumed orders?
```
Comment (antonyshchenko) [inline sequencer/src/main/kotlin/co/chainring/sequencer/core/OrderBook.kt:158]: Would be nice to have I would stay, but not a blocker for merging this

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/OrderBook.kt [hunk 2]
```diff
@@ -96,6 +96,11 @@ class OrderBookLevel(val levelIx: Int, var side: BookSide, val price: BigDecimal
     fun fromCheckpoint(checkpoint: MarketCheckpoint.OrderBookLevel) {
         orderHead = checkpoint.orderHead
         orderTail = checkpoint.orderTail
+        side = when (checkpoint.side) {
+            MarketCheckpoint.BookSide.Buy -> BookSide.Buy
+            MarketCheckpoint.BookSide.Sell -> BookSide.Sell
+            else -> throw IllegalStateException("Unexpected level book side '${checkpoint.side}'")
+        }
```
Comment (antonyshchenko) [inline sequencer/src/main/kotlin/co/chainring/sequencer/core/OrderBook.kt:103]: That's what was causing the issues we've observed in demo

File: sequencer/src/main/kotlin/co/chainring/sequencer/apps/SequencerApp.kt [hunk 3]
```diff
@@ -46,6 +47,7 @@ class SequencerApp(
     val outputQueue: RollingChronicleQueue = defaultOutputQueue,
     val checkpointsPath: Path?,
     val inSandboxMode: Boolean = System.getenv("SANDBOX_MODE").toBoolean(),
+    private val strictReplayValidation: Boolean = System.getenv("STRICT_REPLAY_VALIDATION").toBoolean(),
```
Comment (antonyshchenko) [inline sequencer/src/main/kotlin/co/chainring/sequencer/apps/SequencerApp.kt:50]: 👍

**Review by antonyshchenko — APPROVED** [review]
Nice!

File: sequencer/src/main/kotlin/co/chainring/sequencer/core/OrderBook.kt (hunk 1 above)
Comment (imykolenko) [reply sequencer/src/main/kotlin/co/chainring/sequencer/core/OrderBook.kt:158]: Agreed. Also going to add few more tests

Comment (bholzman) [inline sequencer/src/main/kotlin/co/chainring/sequencer/core/OrderBook.kt:158]: suggest creating a JIRA ticket in the backlog for this and referencing the ticket number in the code, .e.g:
```
// TODO: CHAIN-XXX - also reset consumed orders
```

**Review by bholzman — APPROVED** [review]

**Pull request merged**